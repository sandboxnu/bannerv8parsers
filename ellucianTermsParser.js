/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import moment from 'moment';
import cheerio from 'cheerio';

import cache from '../../cache';
import macros from '../../../macros';
import Request from '../../request';


import EllucianBaseParser from './ellucianBaseParser';
import ellucianSubjectParser from './ellucianSubjectParser';
import { getBaseHost } from '../../employees/util';


const request = new Request('EllucianTermsParser');

class EllucianTermsParser extends EllucianBaseParser.EllucianBaseParser {
  supportsPage(url) {
    return url.includes('bwckschd.p_disp_dyn_sched');
  }


  //
  async main(url) {
    // Possibly load from DEV
    if (macros.DEV && require.main !== module) {
      const devData = await cache.get(macros.DEV_DATA_DIR, this.constructor.name, url);
      if (devData) {
        return devData;
      }
    }

    const resp = await request.get(url);

    const termsAndPostUrl = this.parse(resp.body, url);

    const postUrl = termsAndPostUrl.postUrl;
    let terms = termsAndPostUrl.terms;

    terms = await this.addSubjects(terms, postUrl);


    // Possibly save to dev
    if (macros.DEV && require.main !== module) {
      await cache.set(macros.DEV_DATA_DIR, this.constructor.name, url, terms);

      // Don't log anything because there would just be too much logging.
    }

    return terms;
  }


  minYear() {
    return moment().year();
  }

  isValidTerm(termId, text) {
    const year = text.match(/\d{4}/);
    const minYear = this.minYear();

    if (!year) {
      macros.log('warning: could not find year for ', text);

      //if the termId starts with the >= current year, then go
      const idYear = parseInt(termId.slice(0, 4), 10);

      //if first 4 numbers of id are within 3 years of the year that it was 4 months ago
      if (idYear + 3 > minYear && idYear - 3 < minYear) {
        return true;
      }

      return false;
    }

    //skip past years
    if (parseInt(year, 10) < minYear) {
      return false;
    }
    return true;
  }


  parse(body, url) {
    const formData = this.parseTermsPage(body, url);
    const terms = [];

    formData.requestsData.forEach((singleRequestPayload) => {
      //record all the terms and their id's
      singleRequestPayload.forEach((payloadVar) => {
        if (this.shouldParseEntry(payloadVar)) {
          terms.push({
            termId: payloadVar.value,
            text: payloadVar.text,
          });
        }
      });
    });

    if (terms.length === 0) {
      macros.log('ERROR, found 0 terms??', url);
    }

    const host = getBaseHost(url);

    for (const term of terms) {
      // If this is a term that matches a term in staticHosts
      // Remove

      term.host = host;

      if (term.host === 'neu.edu') {
        if (term.text.toLowerCase().includes(' law ')) {
          term.subCollegeName = 'LAW';
          term.text = term.text.replace(/LAW/gi, '');
        } else if (term.text.toLowerCase().includes(' cps ')) {
          term.subCollegeName = 'CPS';
          term.text = term.text.replace(/CPS/gi, '');
        } else {
          term.text = term.text.replace(/Semester/gi, '');
        }

        term.text = term.text.replace(/\s+/gi, ' ').trim();
      }
    }

    const outputTerms = [];

    // Keep just some of the properties and normalize the data structure to the labeled format.
    for (const term of terms) {
      outputTerms.push({
        type: 'terms',
        value: term,
        deps: null,
      });
    }

    return {
      terms: outputTerms,
      postUrl: formData.postURL,
    };
  }


  shouldParseEntry(entry) {
    if (entry.name === 'p_term') {
      return true;
    }

    return false;
  }


  async addSubjects(terms, postURL) {
    const promises = [];

    terms.forEach((term) => {
      macros.log('Parsing term: ', JSON.stringify(term.value));

      const promise = ellucianSubjectParser.main(postURL, term.value.termId).then((subjects) => {
        term.deps = subjects;
      });

      promises.push(promise);
    });


    // Wait for all the subjects to be parsed.
    await Promise.all(promises);

    return terms;
  }


  //step 1, select the terms
  //starting url is the terms page
  parseTermsPage(body, url) {
    // Parse the dom
    const $ = cheerio.load(body);

    const parsedForm = this.parseForm(url, $('body')[0]);

    if (!parsedForm) {
      macros.error('default form data failed');
      return null;
    }

    const defaultFormData = parsedForm.payloads;


    //find the term entry and all the other entries
    let termEntry;
    const otherEntries = [];
    defaultFormData.forEach((entry) => {
      if (this.shouldParseEntry(entry)) {
        if (termEntry) {
          macros.error('Already and entry???', termEntry);
        }
        termEntry = entry;
      } else {
        otherEntries.push(entry);
      }
    });

    if (!termEntry) {
      macros.error('Could not find an entry!', url, JSON.stringify(parsedForm));
      return null;
    }

    const requestsData = [];

    //setup an indidual request for each valid entry on the form - includes the term entry and all other other entries
    termEntry.alts.forEach((entry) => {
      if (!this.shouldParseEntry(entry)) {
        macros.log('ERROR: entry was alt of term entry but not same name?', entry);
        return;
      }
      entry.text = entry.text.trim();

      if (entry.text.toLowerCase() === 'none') {
        return;
      }
      entry.text = entry.text.replace(/\(view only\)/gi, '').trim();

      entry.text = entry.text.replace(/summer i$/gi, 'Summer 1').replace(/summer ii$/gi, 'Summer 2');

      //dont process this element on error
      if (entry.text.length < 2) {
        macros.log('warning: empty entry.text on form?', entry, url);
        return;
      }

      if (!this.isValidTerm(entry.value, entry.text)) {
        return;
      }


      const fullRequestData = otherEntries.slice(0);

      fullRequestData.push({
        name: entry.name,
        value: entry.value,
        text: entry.text,
      });

      requestsData.push(fullRequestData);
    });

    return {
      postURL: parsedForm.postURL,
      requestsData: requestsData,
    };
  }

  async test() {
    const output = await this.main('https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_dyn_sched');
    for (const thing of output) {
      delete thing.deps;
    }
    macros.log(output);
  }
}


EllucianTermsParser.prototype.EllucianTermsParser = EllucianTermsParser;
const instance = new EllucianTermsParser();


if (require.main === module) {
  instance.test();
}

export default instance;
