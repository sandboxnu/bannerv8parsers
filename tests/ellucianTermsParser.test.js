/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import path from 'path';
import MockDate from 'mockdate';
import fs from 'fs-extra';

import ellucianTermsParser from '../ellucianTermsParser';


beforeAll(() => {
  MockDate.set('Mon Oct 10 2016 00:00:00 -0000');
});

afterAll(() => {
  MockDate.reset();
});


it('has a name', () => {
  //make sure a name is defined
  expect(ellucianTermsParser.name);
});

it('isValidTerm should work', () => {
  expect(ellucianTermsParser.isValidTerm('201630', 'blah blah 2016')).toBe(true);
  expect(ellucianTermsParser.isValidTerm('201630', 'blah blah 2017')).toBe(true);
  expect(ellucianTermsParser.isValidTerm('201630', 'blah blah')).toBe(true);
  expect(ellucianTermsParser.isValidTerm('2016', 'blah blah')).toBe(true);
  expect(ellucianTermsParser.isValidTerm('201', 'blah blah')).toBe(false);
});

it('should behave...', async (done) => {
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianTermsParser', '1.html'), 'utf8');

  const url = 'https://bannerweb.upstate.edu/isis/bwckschd.p_disp_dyn_sched';
  expect(true).toBe(ellucianTermsParser.supportsPage(url));

  expect(ellucianTermsParser.parse(body, url)).toMatchSnapshot();

  done();
});


it('hi there 2', async (done) => {
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianTermsParser', '2.html'), 'utf8');

  const url = 'https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_dyn_sched';

  const output = ellucianTermsParser.parse(body, url);

  expect(ellucianTermsParser.supportsPage(url));

  expect(output).toMatchSnapshot();

  done();
});
