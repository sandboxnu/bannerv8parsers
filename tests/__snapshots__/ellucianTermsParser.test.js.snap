// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hi there 2 1`] = `
Object {
  "postUrl": "https://wl11gp.neu.edu/udcprod8/bwckgens.p_proc_term_date",
  "terms": Array [
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "LAW",
        "termId": "201718",
        "text": "Fall 2016 Quarter",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "CPS",
        "termId": "201715",
        "text": "Fall 2016 Quarter",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "CPS",
        "termId": "201714",
        "text": "Fall 2016 Semester",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "LAW",
        "termId": "201712",
        "text": "Fall 2016 Semester",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "termId": "201710",
        "text": "Fall 2016",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "termId": "201660",
        "text": "Summer 2 2016",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "LAW",
        "termId": "201658",
        "text": "Summer 2016 Quarter",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "CPS",
        "termId": "201655",
        "text": "Summer 2016 Quarter",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "termId": "201652",
        "text": "Summer 2016 Law",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "termId": "201650",
        "text": "Summer Full 2016",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "termId": "201640",
        "text": "Summer 1 2016",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "LAW",
        "termId": "201638",
        "text": "Spring 2016 Quarter",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "CPS",
        "termId": "201635",
        "text": "Spring 2016 Quarter",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "LAW",
        "termId": "201632",
        "text": "Spring 2016 Semester",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "termId": "201630",
        "text": "Spring 2016",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "LAW",
        "termId": "201628",
        "text": "Winter 2016 Quarter",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "neu.edu",
        "subCollegeName": "CPS",
        "termId": "201625",
        "text": "Winter 2016 Quarter",
      },
    },
  ],
}
`;

exports[`should behave... 1`] = `
Object {
  "postUrl": "https://bannerweb.upstate.edu/isis/bwckgens.p_proc_term_date",
  "terms": Array [
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "upstate.edu",
        "termId": "201611",
        "text": "Spring 2017 Summer 2",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "upstate.edu",
        "termId": "201610",
        "text": "Spring 2016",
      },
    },
  ],
}
`;
